lib/howl/interactions/external_command.moon

```diff
@@ -0,0 +1,207 @@
+      directory = @command_line.directory / @opts.path
+    elseif @command_line.directory
+      directory = @command_line.directory
+    elseif @opts.path
+      directory = File @opts.path
+    if not (directory and directory.exists and directory.is_directory)
+      directory = get_cwd!
+    @_chdir directory
+
+  on_update: (text) =>
+    if @list_widget and @list_widget.showing
+      @_update_auto_complete!
+    elseif @auto_show_list
+      if text\umatch '^%s*cd%s+'
+        @_auto_complete_file directories_only: true
+      elseif looks_like_path last_text_part @command_line.text
+        @_auto_complete_file!
+
+  auto_submit: =>
+    @_submit!
+
+  _chdir: (directory) =>
+    @directory = directory
+    trailing = directory.path == File.separator and '' or File.separator
+    @command_line.prompt = markup.howl "\n<operator>[</><directory>#{@directory.short_path}#{trailing}</><operator>] $</> "
+
+  _updir: =>
+    if @directory.parent
+      @_chdir @directory.parent
+
+  _initialize_list_widget: =>
+    @list_widget = ListWidget nil,
+      never_shrink: true
+    @list_widget.max_height = math.floor app.window.allocated_height * 0.5
+    @list_widget.columns =  { {style: 'filename'} }
+    @command_line\add_widget 'completion_list', @list_widget
+
+  _auto_complete_file: (opts={}) =>
+    if not @list_widget
+      @_initialize_list_widget!
+
+    if not @list_widget.showing
+      @list_widget\show!
+
+    if opts.directories_only
+      @directory_reader = (dir) ->
+        dirs = [c for c in *dir.children when c.is_directory]
+        append dirs, 1, dir\join '.'
+        dirs
+    else
+      @directory_reader = (dir) -> dir.children
+
+    @_update_auto_complete!
+
+  _update_auto_complete: =>
+    return unless @list_widget and @list_widget.showing
+    last_part = last_text_part @command_line.text
+
+    if not last_part
+      @list_widget\hide!
+      return
+
+    path, unmatched = parse_path @directory.path .. File.separator .. last_part
+    @list_widget.matcher = file_matcher self.directory_reader(path), path
+    @list_widget\update unmatched
+    @list_widget_path = path
+    @list_widget_unmatched = unmatched
+
+  _auto_complete_command: (text) =>
+    if not @commands
+      @commands = available_commands!
+
+    if @list_widget
+      @list_widget\hide!
+
+    selected = interact.select
+      items: @commands
+      title: 'Commands'
+      columns: { { header: 'Commands', style: 'string' } }
+      submit_on_space: true
+      cancel_on_backspace: true
+      :text
+
+    if selected
+      @command_line\write selected.selection
+      @command_line\write ' '
+
+  _select_completion: =>
+    @list_widget\hide!
+    return unless @list_widget.selection
+
+    filename = @list_widget.selection.name
+    new_path = @list_widget_path / filename
+    @command_line.text = @command_line.text\sub(1, -#@list_widget_unmatched - 1)
+    @command_line\write filename
+    if not new_path.is_directory
```

Comment:
You typically use `if not` in lieu of `unless`, which is fine by me as as I chalk it down to a personal preference, and personally find `unless` to add confusion in some cases. In this case though, `unless new_path.is_directory` read a lot better :)